pkg/segment/sortindex/sortindex.go

```diff
@@ -177,53 +219,122 @@ func AsRRCs(lines []Line, segKeyEncoding uint32) ([]*segutils.RecordResultContai
 	return rrcs, values
 }
 
-func ReadSortIndex(segkey string, cname string, maxRecords int) ([]Line, error) {
+func ReadSortIndex(segkey string, cname string, maxRecordsToRead int) ([]Line, error) {
 	file, err := os.Open(getFilename(segkey, cname))
 	if err != nil {
 		return nil, err
 	}
 	defer file.Close()
 
+	// Read version
+	version := make([]byte, 1)
+	err = binary.Read(file, binary.LittleEndian, &version)
+	if err != nil {
+		return nil, fmt.Errorf("ReadSortIndex: failed reading version: %v", err)
+	}
+	if version[0] != VERSION_SORT_INDEX[0] {
+		return nil, fmt.Errorf("ReadSortIndex: unsupported version: %v", version)
+	}
+
+	// Read Number of unique column values
+	var totalUniqueColValues uint64
+	err = binary.Read(file, binary.LittleEndian, &totalUniqueColValues)
+	if err != nil {
+		return nil, fmt.Errorf("ReadSortIndex: failed reading number of unique column values: %v", err)
+	}
+
 	lines := make([]Line, 0)
 
-	numRecords := 0
-	for numRecords < maxRecords {
-		var valueLen uint16
-		err := binary.Read(file, binary.LittleEndian, &valueLen)
+	numColValues := uint64(0)
+	totalRecordsRead := uint64(0)
+	done := false
+
+	for numColValues < totalUniqueColValues && !done {
+		// Read DType
+		var Dtype segutils.SS_DTYPE
+		err = binary.Read(file, binary.LittleEndian, &Dtype)
 		if err == io.EOF {
 			break
 		}
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("ReadSortIndex: failed reading DType: %v", err)
```

Comment:
the EOF error in line 256 should also be considered as an error and we should return an error to the caller. This is Because we entered the loop since we still had some more colValues to read, but encountered a EOF, and that means the file was prematurely cut.

Reply:
the error statement should say above, and also include the numColValues and totalUnieueColVals var in it